Add config for foreign key constraints

r-ash opened the pull request:
This PR will
* Allow users to run an import using default load step by initialising the import using `dettl("path", default_load = TRUE, db_name = "uat")`
* Improve method naming, to initialise an import use `dettl` instead of `new_import`
* Allow users to configure foreign key constraints for use in `load` stage in the `dettl.yml` config
* Update vignette to include information about how to use this

r-ash commented on tests/testthat/test-run-default-load.R:
```diff
@@ -0,0 +1,53 @@
+                               insert_func(data[2, c("name", "age", "height")]))
+
+  ## Run load with mock INSERT function
+  with_mock("dettl:::insert_data" = mock_insert, {
```
Here I am mocking the code which does the actual insert as that uses an `INSERT INTO RETURNING` pattern which is not available in SQLite. This looked like the best way to work around it though I'm not super happy with it as it results in lots of code in the test and means I'm not testing the entire `get_default_load` function. I wonder if that is just something we have to accept as the test driver is using SQLite instead of Postgres so there will be some differences. 

I'm hoping the explanation lines 15-22 when I start writing the mocking is enough to make this clear too?

richfitz replied:
I'm confused why you have both `mockery`'s mock and testthat's `with_mock` here

r-ash replied:
This was following the examples from `mockery` documentation for `mock` and the "simple integration" section mockery [vignette](https://cran.r-project.org/web/packages/mockery/vignettes/mocks-and-testthat.html). Is there a mockery equivalent or `with_mock`?

I could avoid using the `mockery::mock` as I am not doing any interaction testing but I think that would make the mock function more complicated.

richfitz commented on R/dettl_runner.R:
```diff
@@ -1,24 +1,31 @@
 #' Create an import object using functions defined at specified path
 #'
 #' @param path Path to directory containing functions for import process.
+#' @param default_load If true then load step will use default implementation.
```
I am not sure why this is an argument - this should be a property of the configuration, not the evaluation in my mind?

richfitz commented on R/dettl_runner.R:
```diff
@@ -1,24 +1,31 @@
 #' to using the first db configured.
 #'
-#' @return A dataImport object.
+#' @return A DataImport object.
```
We should document up somewhere how we feel about class names and the like

r-ash replied:
So I updated this after seeing Hadley Wickham's recommendation in Advanced R
> By convention, R6 classes have UpperCamelCase names

Where you thinking documenting this at a repo level or something more general?

richfitz replied:
more generally - we don't have a style guide we can point to yet

richfitz commented on tests/testthat/test-dettl-runner.R:
```diff
@@ -13,11 +13,11 @@ test_that("dettl works as expected", {
   ## object has been created
   expect_false(is.null(import))
-  expect_equal(class(import), c("dataImport", "R6"))
+  expect_equal(class(import), c("DataImport", "R6"))
```
you can also use `expect_is` here to to check class attributes

richfitz commented on tests/testthat/test-foreign-key-constraints.R:
```diff
@@ -0,0 +1,72 @@
+context("foreign-key-constraints")
```
Putting the comment here because I don't know where it would otherwise naturally go - as Alex found out the other day, one has to opt in to SQLite's FK enforcement.  See

https://github.com/vimc/orderly/blob/master/R/db2.R#L568-L572

which we use when creating the connection

r-ash replied:
I've pushed https://github.com/vimc/dettl/pull/10/commits/6839705aae34476d2e9454342aa2f153063a08c3 to address this

richfitz commented on tests/testthat/test-run-default-load.R:
```diff
@@ -0,0 +1,53 @@
+  import <- dettl("example_default_load/", default_load = TRUE,
+                  db_name = "test")
+
+  ## Setup mock INSERT query for testing. This is needed because in real
```
I think that a possibly better solution would be to abstract this bit of logic into a function `insert_into_returning` (or whatever!) that switches based on the SQL dialect.  So if you generalise the `insert_values_into` function so that the form you have is for postgres, then add another one for sqlite (you could do this with either a switch or S3 methods perhaps)

richfitz commented on R/default_load.R:
```diff
@@ -0,0 +1,117 @@
+## impl and test
```
I think a high level overview of the approach here in a comment would be useful

r-ash commented:
Okay I think this addresses all the points raised, in particular
* Now flag for using default load is in the `dettl.yml` config
* Add new generic function for inserting values in DB returning id of added data and impl for SQLite and Postgres
* Better documentation

I've created a separate ticket for moving away from having to configure the foreign key constraints and instead reading them from the DB.

richfitz commented on R/db.R:
```diff
@@ -7,11 +7,28 @@
+#' enabled.
+#'
+#' @keywords internal
+dbi_db_connect <- function(drv, ...) {
```
This way is fine, but it took me a minute to work out what is going on. I think the way we do it in `orderly` is closer to 

```r
db_connect <- function(type, path) {
  x <- dettl_db_args(path, type)
  con <- do.call(DBI::dbConnect,
                 c(list(x$driver()), x$args))
  sqlite_enable_fk(con)
  con
}

sqlite_enable_fk <- function(con) {
  if (class(drv) == "SQLiteDriver") {
    ## Foreign key constraints aren't enabled by default in SQLite. Ensure
    ## they are enabled each time we connect to the db.
    DBI::dbExecute(con, "PRAGMA foreign_keys = ON")
  }
}
```

which I think is slightly easier to follow

r-ash replied:
So I did it this way because the current `dbi_db_connect` method is used in several of my tests to connect to test db. I could move that connect for testing into a helper function, perhaps that will be clearer.

richfitz replied:
it can always be done later if you don't think it's worth it - it's up to you

r-ash replied:
No that's good, I've updated that now.

richfitz commented on R/default_load.R:
```diff
@@ -0,0 +1,81 @@
+}
+
+strip_primary_key_column <- function(data, primary_key) {
+  column_names <- !(names(data) == primary_key)
```
`<- names(data) != primary_key`

?

r-ash replied:
Silly mistake, I've fixed that

richfitz commented on tests/testthat/example_default_load/dettl.yml:
```diff
@@ -0,0 +1,24 @@
+  verification_queries: test_queries
+  test: R/test_load.R
+
+default_load: TRUE
```
should this go into the `load` block as a key `default: true`?

richfitz replied:
and I think the config validation needs a  check that if the default load is given then `load: func` is missing or empty

richfitz commented on README.md:
```diff
@@ -18,3 +18,7 @@ To set these create a file called `envir.yml` in the root of the directory which
+
+## Running tests which rely on PostgreSQL database
+
+Most tests run against a local SQLite database but some require a PostgreSQL database. These tests will be skipped by default when running locally. To enable them run the `scripts/test_db_start.sh` script to run up a Postgres db within a docker image. When finished testing locally run `scripts/test_db_stop.sh` to stop the image.
```
To enable them run the `scripts/test_db_start.sh` script to run up a Postgres db within a docker ~image~ container. When finished testing locally run `scripts/test_db_stop.sh` to stop the ~image~ container.

richfitz replied:
(an image is like a class, a container is like an instance)

r-ash replied:
Thanks, fixed this too.

richfitz commented on R/db.R:
```diff
@@ -9,9 +13,22 @@ db_connect <- function(type, path) {
+#'
+#' @keywords internal
+sqlite_enable_fk <- function(con) {
+  if (class(con) == "SQLiteConnection") {
```
Can you please update this to be `inherits(con, "SQLiteConnection")` - I should have picked that up before but that's the only safe way to ask about class membership (class vectors can be any length which makes `==` work element-wise)

richfitz commented on R/dettl_runner.R:
```diff
@@ -30,20 +36,20 @@ new_import <- function(path, db_name = NULL) {
 #'
 run_import <- function(import, run_stages = NULL, dry_run = FALSE) {
-  if (is.null(import) || class(import) != "dataImport") {
+  if (is.null(import) || class(import) != "DataImport") {
```
as above, `class(import) != "DataImport"` should be `!inherits(import, "DataImport")`
